### PR TITLE
Query class

### PR DESCRIPTION
**Pull request by AverageHelper**
# Description
We introduce here the `Query` class, where many of the collection query code lives.

These bits will be important when we introduce `CollectionReference` and `DocumentReference` as distinct classes for subcollection mocking.

## Related issues
Depends on changes in #36 

## How to test
```
jest
```

### CODE REVIEW

File: mocks/firestore.js [hunk 1]
```diff
@@ -177,21 +150,28 @@ class FakeFirestore {
   }
 
   orderBy() {
-    mockOrderBy(...arguments);
-    return this;
+    return this.query.orderBy(...arguments);
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/firestore.js:153]: Actually, we can! `FakeFirestore.orderBy` calls `Query.orderBy`, which returns the same query instance (or whatever injected value that folks want it to return instead). Chaining still works.

File: mocks/auth.js [hunk 2]
```diff
@@ -4,9 +4,12 @@ const mockSendVerificationEmail = jest.fn();
 const mockSignInWithEmailAndPassword = jest.fn();
 const mockSendPasswordResetEmail = jest.fn();
 const mockVerifyIdToken = jest.fn();
+const mockGetUser = jest.fn();
+const mockSetCustomUserClaims = jest.fn();
 
 class FakeAuth {
   constructor(currentUser) {
+    currentUser = currentUser || {};
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/auth.js:12]: Probably. The idea was mainly to preserve some semblance of type-safety. If someone constructs it with an empty string or `false`, we'd probably want that to be an empty object instead. I'm not sure it matters a whole lot. I'll go ahead and move it in.

File: mocks/firestore.js [hunk 3]
```diff
@@ -24,48 +19,29 @@ const mockBatchCommit = jest.fn();
...
-      return docs.forEach(callback);
-    },
-  };
-}
-
 class FakeFirestore {
   constructor(stubbedDatabase = {}) {
     this.isFetchingSingle = false;
     this.database = stubbedDatabase;
+    this.query = new query.Query('', this);
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/firestore.js:35]: Can do!

File: mocks/query.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  constructor(collectionName, firestore) {
+    this.collectionName = collectionName;
+    this.firestore = firestore;
+  }
+
+  get() {
+    mockGet(...arguments);
+    // Use DFS to find all records in collections that match collectionName
+    const requestedRecords = [];
+    // requestedRecords.push(...this.firestore.database[this.collectionName]);
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/query.js:21]: This line? This file? ...?

File: mocks/firestore.js [hunk 5]
```diff
@@ -82,15 +58,7 @@ class FakeFirestore {
...
   }
 
-  where() {
-    this.isFetchingSingle = false;
-    mockWhere(...arguments);
-    return this;
-  }
-
   get() {
-    mockGet(...arguments);
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/firestore.js:92]: That makes sense. Added back in.

File: mocks/query.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbatson5) [reply mocks/query.js]: Sorry -- the commented out line

[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/query.js]: Ah yes. Done.

**Comment by sbatson5** [comment]
Sorry for the delay in getting to reviewing this. Looks good to merge to me 👍

**Comment by AverageHelper** [comment]
Thanks! #38 is ready for review

**Pull request merged**